### DISCUSSION TITLE
Bug ?

### DISCUSSION THREAD

**Discussion by ziki17** [fnando/i18n-js#664] [Q&A]
Facing this error: Sprockets::FileNotFound: couldn't find file 'i18n.js' with type 'application/javascript' Agter upgrading to Rails 7.0.1, Ruby 3.0.6.
I have tried both: gem 'i18n-js', ">= 4.0" as well as gem 'i18n-js'.
Included all three
//= require i18n
//= require i18n.js
//= require i18n/translations

This works fine with rails 6.1

**Comment by fnando** [comment]
Can you please create a sample app that reproduce this error?

**Reply by far-wtag** [reply]
Hi @ziki17 , @fnando is there any update on this? I am currently working in an app which I have recently upgrade with Rails 7 and after that i18n-js gem is not working anymore. I have checked this [MIGRATING_FROM_V3_TO_V4](https://github.com/fnando/i18n-js/blob/main/MIGRATING_FROM_V3_TO_V4.md) documentation, but this doesn't seems work for me.

The error I am getting is, `Uncaught ReferenceError: I18n is not defined` in my server console.
Is there any easy way/suggestions to setup this https://github.com/fnando/i18n-js for a ROR application?

**Rails 7.0
Ruby 3.2.2**

### Gemfile
```ruby
gem 'i18n-js'
```

### app/assets/javascripts/i18n.js

```ruby
import { I18n } from 'i18n-js';
import translations from '/../config/locales/';

const userLocale = document.documentElement.lang;

const i18n = new I18n();
i18n.store(translations);
i18n.defaultLocale = 'en';
i18n.enableFallback = true;
i18n.locale = userLocale;
window.I18n = i18n;

export { I18n };
```

### app/assets/javascripts/application.js

```ruby
//= require i18n
//= require_tree .
```

### application.html.erb

```ruby
<script type="text/javascript">
    I18n.defaultLocale = "<%= I18n.default_locale %>";
    I18n.locale = "<%= I18n.locale %>";
</script>
```

this is what I have configured as per the documentations. Let me know if there is anything else need as well since I have very limited understanding on **Rails 7** changes formats.

**Reply by fnando** [reply]
@far-wtag the first thing I'm noticing is that your `application.html.erb` is wrong. There's no global `I18n` anymore, so you need to either expose the `i18n` constant you created on `app/assets/javascripts/i18n.js` with something like `window.I18n = i18n`, or read the locale out of your html (usually that's how I do it). You can, for instance, set these values using `data` attributes, then reading them.

```erb
<html data-default-locale="<%= I18n.default_locale %>" lang="<%= I18n.locale %>">
  <!-- your content here -->
</html>
```

Then from your `app/assets/javascripts/i18n.js` file you can use `document.documentElement.dataset.defaultLocale` and `document.documentElement.lang`.

**Reply by far-wtag** [reply]
Thanks for the reply @fnando . According to your suggestion, I have updated the below two files and still getting errors. Since. I am not very familiar with JS thats why not sure whats wrong there. Is there anything else needs to setup?

```ruby
# config/locales
  # de.yml
     - have translations in german
  # en.yml
     - have translations in english   
```

```ruby
# app/assets/javascripts/i18n.js

const { I18n } = require ('i18n-js')
const { translations } = require ('/../config/locales/')

// Fetch user locale from html#lang.
// This value is being set on `app/views/layouts/application.html.erb` and
// is inferred from `ACCEPT-LANGUAGE` header.
const userLocale = document.documentElement.lang;
const defaultLocale = document.documentElement.dataset.defaultLocale;

const i18n = new I18n();
i18n.store(translations);
i18n.defaultLocale = defaultLocale || 'en';
i18n.enableFallback = true;
i18n.locale = userLocale;
window.I18n = i18n;

module.exports = { I18n };
```

```ruby
# application.html.erb

<html data-default-locale="<%= I18n.default_locale %>" lang="<%= I18n.locale %>">
  <head>..
```

The error i am getting is, `Uncaught ReferenceError: require is not defined at i18n.source.js:3:1` and for this line `const { I18n } = require ('i18n-js')`

**Reply by fnando** [reply]
@far-wtag now the error is something related with your build pipeline, and can't really help you with that, because I don't know what you're using (rollup, webpack, esbuild, something else?) and how things are set up.

**Reply by far-wtag** [reply]
@fnando I am using the old sprocket asset pipeline. And recently just upgraded the ruby and rails version. With older version rails 5.2, it was working.

**Reply by fnando** [reply]
it was working because you probably used i18n-js v3, which doesn't use modern tooling and was relying solely on sprockets. I recommend that you take a look at esbuild/rollup, which is the way to go with Rails 7.

**Reply by far-wtag** [reply]
>it was working because you probably used i18n-js v3

@fnando Yes, it was using v3
